File: TrolleyCore/API Connections/TRLNetworkManager.m
//
//  TRLNetworkManager.m
//  TRLNetwork
//
//  Created by Harry Wright on 13.09.17.
//  Copyright © 2017 Off-Piste.
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in all
//  copies or substantial portions of the Software.
//
//   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//  SOFTWARE.
//

#import "TRLNetworkManager.h"
#import "TRLNetworkManager_Options.h"
#import "TRLNetworkManager_Private.h"

#import "TRLNetwork_Private.h"
#import "TRLParsedURL.h"

#import "TRLURLEncoding.h"

#import "Swift-Fixed-Header.h"

@implementation TRLNetworkManager

- (instancetype)initWithURL:(NSString *)url APIKey:(NSString *)key {
    if (self = [super init]) {
        self->_network = [[TRLNetwork alloc] initWithURLString:url];
        [_network.parsedURL addPath:key];
    }
    return self;
}

- (instancetype)initWithOptions:(TRLOptions *)options {
    return [self initWithURL:options.url APIKey:options.key];
}

- (TRLRequest *)get:(NSString *)path {
    return [self get:path parameters:NULL];
}

- (TRLRequest *)get:(NSString *)path
         parameters:(Parameters *)parameters
{
    TRLURLEncoding *encoding = [TRLURLEncoding defaultEncoding];
    return [self get:path parameters:parameters encoding:encoding];
}

- (TRLRequest *)get:(NSString *)path
         parameters:(Parameters *)parameters
           encoding:(id<TRLURLParameterEncoding>)encoding
{
    return [self get:path parameters:parameters encoding:encoding headers:NULL];
}

- (TRLRequest *)get:(NSString *)path
         parameters:(Parameters *)parameters
           encoding:(id<TRLURLParameterEncoding>)encoding
            headers:(HTTPHeaders *)headers
{
    return [_network get:path parameters:parameters encoding:encoding headers:headers];
}

- (NSString *)description {
    return _network.description;
}

@end

File: Trolley/Database/TRLNetworkManager.swift
//
//  NetworkManager.swift
//  Network
//
//  Created by Harry Wright on 19.06.17.
//  Copyright © 2017 Trolley. All rights reserved.
//

import Foundation
import Alamofire

let kLocalURL: String = "http://localhost:8080/API"
let kLocalIPURL: String = "http://127.0.0.1:8080/API"

extension String {
    
    static func urlRoute(for items: String...) -> String {
        return items.joined(separator: "/")
    }
    
}

public struct TRLNetworkManager {
    
    var network: TRLNetwork
    
    internal init(network: TRLNetwork, key: String) {
        self.network = network
        self.network.parsedURL._addPath(key)
    }
    
    internal init(_ url: URLConvertible, key: String) throws {
        self.network = try TRLNetwork(url)
        self.network.parsedURL._addPath(key)
    }
    
    internal init?(url: URLConvertible, key: String) {
        do {
            self.network = try TRLNetwork(url)
            self.network.parsedURL._addPath(key)
        } catch {
            return nil
        }
    }
    
}

extension TRLNetworkManager : CustomStringConvertible {
    
    public var description: String {
        return self.network.parsedURL.description
    }
    
}

public extension TRLNetworkManager {
    
    func get(
        _ route: String,
        with parameters: Parameters? = nil,
        encoding: ParameterEncoding = URLEncoding.default,
        headers: HTTPHeaders? = nil
        ) -> TRLRequest
    {
        return self.network.get(route, with: parameters, encoding: encoding, headers: headers)
    }
    
    func get(
        item: String,
        in route: String,
        with parameters: Parameters? = nil,
        encoding: ParameterEncoding = URLEncoding.default,
        headers: HTTPHeaders? = nil
        ) -> TRLRequest
    {
        let route = String.urlRoute(for: route, item)
        return self.network.get(route, with: parameters, encoding: encoding, headers: headers)
    }
    
}

// TODO: Make sure this works

public extension TRLNetworkManager {
    
    func post(
        _ object: Object,
        in route: String,
        with parameters: Parameters = [:],
        headers: HTTPHeaders? = [:]
        ) -> TRLRequest
    {
        return self.network.post(route, with: parameters, encoding: object, headers: headers)
    }
    
}
